### ISSUE TITLE
[META][UX][DX] Improve the form validation and behavior of machine_name fields.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1609]
This "bug" most likely applies to other forms as well, but I found it while testing things out with form validation in the layout add form (`admin/structure/layouts/add`)...
1. Head to that form and without entering any value in any field or selecting any template, scroll down to the form submit buttons and hit the "Create layout" button. You get a series of validation errors including one for the machine name and that field is presented to you in edit mode with red border:
   
   ![backdrop-layout_add_form-validation_errors_with_create_button-machine_name_field_exposed_in_edit_mode](https://cloud.githubusercontent.com/assets/2423362/12946721/9a5585c4-d04a-11e5-8b3e-96e5c0b31db8.png)
   
   ...at this point, the user looses the auto-generation functionality of the machine name and there is no way to get it back (other than a page refresh of course but that is not the point).
2. Now, refresh the page (so you get rid of the validation error messages) but instead of hitting the "Create layout" button, click the "+ Add visibility condition" link (still not filling in any fields in order to trigger the validation errors). Because of the fact that this link does not cause a page refresh (#1606), you'll have to scroll back up to the top of the page in order to see the validation errors. Although you get the exact 4 same errors, this time the machine name field is not visible in edit mode (although we still throw a validation error about it) and there is no red border around the layout name and path fields:
   
   ![backdrop-layout_add_form-validation_errors_with_add_visibility_condition_link-machine_name_field_not_exposed](https://cloud.githubusercontent.com/assets/2423362/12946726/a6d6a62a-d04a-11e5-8032-ec0ab31171d2.png)

I believe that:
- we should not trigger the machine name field at all in this case and also hide the validation error about the machine name from the user (they will fill in the layout name field anyways).
- the validation of this field should only be triggered if the user has clicked the "edit" link in order to manually enter a machine name.
- the validation errors when the user hits the "+ Add visibility condition" link should be accompanied by red borders around the missing required fields.
- #1608

**Comment by BWPanda** [comment]
Yes, I think validation of machine names should be ignored as long as the field they're linked to is empty. If the linked field has a value, or the machine name field is editable/visible, then validation can be run on it.
